src/nvim/eval/funcs.c

```diff
@@ -601,13 +601,8 @@ static void f_bufname(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   rettv->vval.v_string = NULL;
   if (argvars[0].v_type == VAR_UNKNOWN) {
     buf = curbuf;
-  } else {
-    if (!tv_check_str_or_nr(&argvars[0])) {
-      return;
-    }
-    emsg_off++;
-    buf = tv_get_buf(&argvars[0], false);
-    emsg_off--;
+  } else if (!(buf = tv_get_buf_from_arg(&argvars[0]))) {
+    return;
```

Comment:
There's no need for an early return when `buf` is set to `NULL`.

Reply:
Sorry, I meant to mention that in the PR post, not just the commit message.

Btw I'm happy to address all the requested changes to this PR if it'll help. 😃